api/v1/__init__.py

```diff
@@ -0,0 +1,180 @@
+from pathlib import Path
+from typing import Any, Dict, Tuple
+
+import requests
+import structlog
+import yaml
+from fastapi import APIRouter, BackgroundTasks, HTTPException
+from git.repo import Repo
+from jsonschema import validate
+from jsonschema.exceptions import ValidationError
+from sqlmodel import Session
+
+from apps.backport.datasync.datasync import upload_gzip_dict
+from etl import config, paths
+from etl import grapher_model as gm
+from etl.command import main as etl_main
+from etl.db import get_engine
+from etl.files import yaml_dump
+from etl.helpers import read_json_schema
+from etl.metadata_export import merge_or_create_yaml, reorder_fields
+from etl.paths import SCHEMAS_DIR
+
+from .. import utils
+from .schemas import Indicator, UpdateIndicatorRequest
+
+log = structlog.get_logger()
+
+
+engine = get_engine()
+
+DATASET_SCHEMA = read_json_schema(path=SCHEMAS_DIR / "dataset-schema.json")
+
+
+v1 = APIRouter()
+
+
+@v1.get("/v1/health")
+def health() -> dict:
+    return {"status": "ok"}
+
+
+@v1.put("/v1/indicators/{indicator_id}")
+def update_indicator(indicator_id: int, update_request: UpdateIndicatorRequest, background_tasks: BackgroundTasks):
+    _validate_data_api_url(update_request.dataApiUrl)
+
+    # update YAML file
+    with Session(engine) as session:
+        db_indicator = gm.Variable.load_variable(session, indicator_id)
+
+    if db_indicator.catalogPath is None:
+        raise HTTPException(403, "Only indicators from the ETL can be edited. Contact us if you really need this.")
+
+    override_yml_path, table_name, indicator_short_name = _parse_catalog_path(str(db_indicator.catalogPath))
+
+    # convert incoming indicator updates into dictionary
+    meta_dict = _indicator_metadata_dict(update_request.indicator, table_name, indicator_short_name)
+
+    yaml_str = merge_or_create_yaml(
+        meta_dict,
+        override_yml_path,
+    )
+
+    # reorder YAML and dump it back again in nice format
+    yaml_str = yaml_dump(reorder_fields(yaml.safe_load(yaml_str)))
+    assert yaml_str
+
+    # validate YAML against our schema
+    try:
+        validate(instance=yaml.safe_load(yaml_str), schema=DATASET_SCHEMA)
+    except ValidationError as e:
+        raise HTTPException(status_code=422, detail=e.message)
+
+    # NOTE: the order of all updating "tracks" is random, we can rethink it if we find
+    # ourselves running into race conditions.
+    if not update_request.dryRun:
+        with open(override_yml_path, "w") as f:
+            f.write(yaml_str)
+
+    if not update_request.dryRun:
+        _update_indicator_in_r2(db_indicator, update_request.indicator)
+
+    if config.ETL_API_COMMIT:
+        background_tasks.add_task(_commit_and_push, override_yml_path, ":robot: Metadata update by Admin")
```

Comment:
It's not clear to me that this should be a background task. The whole request should probably fail if you can't commit the result and push it, perhaps it should even return the repo to a known good state on error so that the next attempt to edit is likely to succeed.

Reply:
The commit should also happen before a push to R2, since we want to avoid pushing to R2 and a commit+push failing. The R2 push can be in the background.

Reply:
> The commit should also happen before a push to R2

Yep, good point!

> perhaps it should even return the repo to a known good state on error so that the next attempt to edit is likely to succeed.

I'm not sure yet what a bad state would look like (which is why I'm logging all requests to slack, to find it out), so I'd rather implement it when this becomes a problem.